django/db/models/deletion.py

```diff
@@ -270,6 +270,57 @@ def delete(self):
         # number of objects deleted for each model label
         deleted_counter = Counter()
 
+        # caches query results of signals to avoid one hit per ['pre_', 'post_']
+        cached_pk_sets = dict()
+
+        def send_m2m_changed(model, obj, action):
+            """
+            Sends `m2m_changed` to all m2m relationships of this `model` that
+            contain at least one row in the table.
+            """
+            m2m_relations = [
+                f for f in model._meta.get_fields(include_hidden=True)
+                if f.many_to_many
+            ]
+
+            from django.db.models import ManyToManyField, ManyToManyRel
+            # for each relation, get all ids of the other side of
+            # the relation and send the signal with them
+            for m2m_relation in m2m_relations:
+
+                # fixme: find a simpler idiom to select the manager and sender.
+                # ATM is with isinstance and getattr(obj, attr_name):
+                if isinstance(m2m_relation, ManyToManyField):
+                    attr_name = m2m_relation.name
+                    sender = m2m_relation.remote_field.through
+                elif isinstance(m2m_relation, ManyToManyRel):
+                    attr_name = m2m_relation.get_accessor_name()
+                    sender = m2m_relation.through
+                else:
+                    # other many_to_many fields are ignored.
+                    continue
+
+                # don't send if no one is listening to avoid a db hit
+                if not signals.m2m_changed.has_listeners(sender):
+                    continue
+
+                key = (obj, attr_name)
+                if key not in cached_pk_sets:
+                    manager = getattr(obj, attr_name)
+
+                    # hit db and cache the results
+                    cached_pk_sets[key] = \
```

Comment:
use a single line here -- lines up to 119 characters are preferred when it helps readability.